lib/zip_source_buffer.c

```diff
@@ -429,30 +429,32 @@ static bool
 buffer_grow_fragments(buffer_t *buffer, zip_uint64_t capacity, zip_error_t *error) {
     zip_buffer_fragment_t *fragments;
     zip_uint64_t *offsets;
+    int result;
+    zip_uint64_t tmp_fragments_capacity, tmp_offsets_capacity, additional_elements;
 
-    if (capacity < buffer->fragments_capacity) {
+    if (capacity <= buffer->fragments_capacity) {
         return true;
     }
 
-    zip_uint64_t fragments_size = sizeof(buffer->fragments[0]) * capacity;
-    zip_uint64_t offsets_size = sizeof(buffer->fragment_offsets[0]) * (capacity + 1);
-
-    if (capacity == ZIP_UINT64_MAX || fragments_size < capacity || fragments_size > SIZE_MAX|| offsets_size < capacity || offsets_size > SIZE_MAX) {
-        zip_error_set(error, ZIP_ER_MEMORY, 0);
+    tmp_fragments_capacity = buffer->fragments_capacity;
+    additional_elements = capacity - buffer->fragments_capacity;
+    result = _zip_realloc(&buffer->fragments, &tmp_fragments_capacity, sizeof(buffer->fragments[0]), additional_elements);
+    if (result != ZIP_ER_OK) {
+        zip_error_set(error, result, 0);
         return false;
     }
 
```

Comment:
Why do you need the temporary variables? Why not passing in pointers to the members of `buffer` directly to `zip_realloc`.

I know that two arrays need to be resized, but each has its own capacity member, and if `zip_realloc` succeed for the first one, that is already resized and doesn't need to be resized again. So it would probably be best to decouple these two sections completely: If `capacity` > `buffer->fragments_capacity`, resize `buffer->fragments`. If `capacity + 1 > buffer->offsets_capacity`, resize `buffer->fragment_offsets`.

Also, `capacity + 1` needs an overflow check.

Reply:
> Why do you need the temporary variables? Why not passing in pointers to the members of `buffer` directly to `zip_realloc`.
> 
> I know that two arrays need to be resized, but each has its own capacity member, and if `zip_realloc` succeed for the first one, that is already resized and doesn't need to be resized again. So it would probably be best to decouple these two sections completely: If `capacity` > `buffer->fragments_capacity`, resize `buffer->fragments`. If `capacity + 1 > buffer->offsets_capacity`, resize `buffer->fragment_offsets`.

The reason here is that those two arrays have actually a single capacity member, `fragments_capacity`. There's no `offsets_capacity`. So I'd like to update `fragments_capacity` after both reallocs succeeded.

> 
> Also, `capacity + 1` needs an overflow check.

Will look into it.

Reply:
I see, I've misread the code, sorry. Your solution is correct.

Reply:
> Also, `capacity + 1` needs an overflow check.

Added a comment about why this overflow is not possible.

Reply:
Ah, I see, I just didn't fully understand the comment. This seems clearer to me, feel free to reword it if you disagree:

This is technically not correct if buffer->fragment_capacity is 0, but results in the correct required new capacity for buffer->fragment_offsets in all cases (one element more than for buffer->fragments). Also, since we added additional_elements (which is 1 or more) to buffer->fragments_capacity in the previous realloc, we know it can't overflow.

Reply:
Used your wording, thanks.